src/main/flight/autopilot.c

```diff
@@ -91,14 +150,282 @@ void altitudeControl(float targetAltitudeCm, float taskIntervalS, float vertical
+    const float pitchVelProp = -cos_approx(normCGERadians); // +1 when drifting backwards, -1 when drifting forwards
+
+    float velocity = gpsSol.groundSpeed;
+    float acceleration = (velocity - posHold.previousVelocity) * gpsDataFreqHz; // positive when moving away
+    posHold.previousVelocity = velocity;
+
+    // include a target based D element. This is smoother and complements groundcourse measurements.
+    float velocityToTarget = (posHold.distanceCm - posHold.previousDistanceCm) * gpsDataFreqHz;
+    posHold.previousDistanceCm = posHold.distanceCm;
+
+    // roll
+    float velocityRoll = rollVelProp * velocity + velocityToTarget * rollProportion;
+    float accelerationRoll = rollVelProp * acceleration;
+    // lowpass filters
+    pt1FilterUpdateCutoff(&velocityRollLpf, pt1Gain);
+    velocityRoll = pt1FilterApply(&velocityRollLpf, velocityRoll);
+    pt2FilterUpdateCutoff(&accelerationRollLpf, pt1Gain);
+    accelerationRoll = pt2FilterApply(&accelerationRollLpf, accelerationRoll);
+    float rollD = velocityRoll * positionPidCoeffs.Kd;
+    float rollA = accelerationRoll * positionPidCoeffs.Kf;
+
+    // pitch
+    float velocityPitch = pitchVelProp * velocity + velocityToTarget * pitchProportion;
+    float accelerationPitch = pitchVelProp * acceleration;
+
+    // lowpass filters
+    pt1FilterUpdateCutoff(&velocityPitchLpf, pt1Gain);
+    velocityPitch = pt1FilterApply(&velocityPitchLpf, velocityPitch);
+    pt2FilterUpdateCutoff(&accelerationPitchLpf, pt1Gain);
+    accelerationPitch = pt2FilterApply(&accelerationPitchLpf, accelerationPitch);
+    float pitchD = velocityPitch * positionPidCoeffs.Kd;
+    float pitchA = accelerationPitch * positionPidCoeffs.Kf;
+
+    // ** test code to handle user yaw inputs ** //
+    if (autopilotConfig()->position_allow_yaw) {
+        // code that we know is effective for fixing user yaw inputs (none at present lol)
+        if (autopilotConfig()->position_test_yaw_fix) {
+
+            // iTerm
+            // Note: accumulated iTerm opposes wind, which is a constant earth-referenced vector
+            // Hence we accumulate earth referenced iTerm
+            // The sign of the iTerm input is determined in relation to the average error bearing angle
+            // Finally the abs value of the accumulated iTerm is proportioned to pitch and roll
+            // Based on the difference in angle between the nose of the quad and the current error bearing
+
+            if (!posHold.isStarting){
+                // only integrate while not actively stopping
+
+                float bearingRadians = bearingDeg * RAD; // 0-360, so constrain to +/- π
+                if (bearingRadians > M_PIf) {
+                    bearingRadians -= 2 * M_PIf;
+                } else if (bearingRadians < -M_PIf) {
+                    bearingRadians += 2 * M_PIf;
+                }
+
+                const float error = posHold.distanceCm * positionPidCoeffs.Ki * gpsDataIntervalS;
```

Comment:
Sorry just trying to understand the logic, why
float acceleration = (velocity - posHold.previousVelocity) * gpsDataFreqHz; 
use time part in Hz, but here it is in seonds? 

Also "error" name is a huge mislead, as I was first thought it was a P calculation.

Reply:
acceleration is derivative of velocity - it is divided by dt
This calculates integral (assuming from Ki coefficient)  - it is multiplied by dt

Reply:
Oh, indeed, thank you.